cargo/src/audit/mod.rs

```diff
@@ -5,3 +5,124 @@
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use quick_xml as xml;
+use std::io;
+use std::path::Path;
+use std::path::PathBuf;
+
+#[allow(unused_imports)]
+use tracing::{debug, error, info, warn, Level};
+
+#[allow(unused_imports)]
+use crate::cli::{Opts, SrcDir, SrcTar};
+
+#[allow(dead_code)]
+const EXCLUDE_ITEMS: &[&str] = &[
+    // These two are excluded because they are fundamentally
+    // silly and can never be fixed.
+    // https://rustsec.org/advisories/RUSTSEC-2020-0071.html
+    // https://rustsec.org/advisories/RUSTSEC-2020-0159.html
+    "RUSTSEC-2020-0071",
+    "RUSTSEC-2020-0159",
+];
+
+pub enum Src {
+    Tar(SrcTar),
+    Dir(SrcDir),
+}
+
+pub fn read_service(p: &Path) -> Result<Src, io::Error> {
+    // use xml::events::BytesStart;
+    use xml::events::Event;
+    use xml::name::QName;
+    use xml::reader::Reader;
+    if let Ok(mut reader) = Reader::from_file(p) {
+        reader.trim_text(true);
+        // let mut skip_buf = Vec::new();
+        let mut operations = Vec::new();
+        let mut is_target_tag = false;
+        let mut is_srctar = false;
+        let mut is_srcdir = false;
+
+        loop {
+            if let Ok(read_event) = reader.read_event_into(&mut operations) {
```

Comment:
Why not just make a Serde type to deserialise into?

Reply:
I was planning to but I think we don't need to. Oh I guess you mean the `serialize` feature. I just mindlessly copied over that line from another project of mine 🥴

Reply:
@firstyear we dont need (de)serialize since we just need to look for the `service` tag with *attribute* `name` that must have the value `cargo_vendor`. 👀

Reply:
I think more that if you did serialise here, you would probably end up with less parsing code. Nothing wrong logically with what you have now, but it may be smaller/clearer.